[CI] DimensionValuesByteRefGroupingAggregatorFunctionTests testSimple failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #11357 / almalinux-8-aarch64_checkpart3_platform-support-arm](https://gradle-enterprise.elastic.co/s/ejhidxpeuxhic)
- [elasticsearch-pull-request #101177 / part-3](https://gradle-enterprise.elastic.co/s/ois6y5li4rhz2)
- [elasticsearch-pull-request #101021 / part-3](https://gradle-enterprise.elastic.co/s/jv7qxinbcwe5y)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:esql:test" --tests "org.elasticsearch.xpack.esql.expression.function.aggregate.DimensionValuesByteRefGroupingAggregatorFunctionTests.testSimple" -Dtests.seed=E7783B22B5F91F8E -Dtests.locale=es-SV -Dtests.timezone=Asia/Beirut -Druntime.java=25
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.esql.expression.function.aggregate.DimensionValuesByteRefGroupingAggregatorFunctionTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testSimple'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: <{513=[[6a 49 65 56 57 46 53 56 33 5a 67 6e 76 51 44 70 4b 72 79 37]], 516=[[52 78 4c 56 48 72 7a 4c 6f 78 38 59 36 4c 41 33 6c 62 68 77]], 519=[], 520=[[50 56 51 4d 49 67 5a 43 62 38 62 59 43 31 49 77 74 68 61 6f], [37 76 4c 33 79 4f 75 68 61 4a 72 31 47 61 58 31 38 77 78 4b], [48 66 74 77 61 4f 54 52 38 53 43 56 72 69 49 6e 43 53 46 39]], 10=[], 522=[[33 44 57 68 75 61 51 48 33 72 75 72 58 45 39 39 75 30 36 6d]], 15=[[6c 7a 4c 4c 79 66 36 52 71 44 71 4a 48 32 70 54 77 57 42 33], [35 4a 44 45 55 79 31 37 4a 32 53 72 5a 68 4b 76 73 7a 4c 79], [35 4b 4a 4f 49 6b 61 78 34 33 30 4a 4f 55 39 31 79 6d 52 6c]], 528=[[53 6c 49 42 6b 51 6f 46 4f 50 47 62 37 35 31 58 49 6e 7a 31]], 18=[[55 52 35 68 65 71 30 30 47 6c 59 6a 39 68 50 4e 48 56 38 6f]], 533=[[43 31 32 44 42 32 50 4d 43 6f 6a 6a 36 4e 61 74 44 41 4c 71]], 534=[[39 59 77 77 36 6e 72 70 79 43 6c 44 32 6c 4e 46 30 43 72 62]], 25=[[76 6b 77 38 54 62 53 47 61 6b 6d 72 54 63 58 4f 71 52 75 4d]], 538=[[7
[truncated]
```

**Issue Reasons:**
- [main] 3 failures in test testSimple (0.7% fail rate in 425 executions)
- [main] 2 failures in step part-3 (0.8% fail rate in 263 executions)
- [main] 2 failures in pipeline elasticsearch-pull-request (0.8% fail rate in 261 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -501,9 +501,6 @@ tests:
 - class: org.elasticsearch.readiness.ReadinessClusterIT
   method: testReadinessDuringRestartsNormalOrder
   issue: https://github.com/elastic/elasticsearch/issues/136955
-- class: org.elasticsearch.xpack.esql.expression.function.aggregate.DimensionValuesByteRefGroupingAggregatorFunctionTests
-  method: testSimple
-  issue: https://github.com/elastic/elasticsearch/issues/137378
 - class: org.elasticsearch.xpack.ilm.TimeSeriesDataStreamsIT
   method: testSearchableSnapshotAction
   issue: https://github.com/elastic/elasticsearch/issues/137167

--- a/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/DimensionValuesByteRefGroupingAggregatorFunction.java
+++ b/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/DimensionValuesByteRefGroupingAggregatorFunction.java
@@ -237,16 +237,17 @@ public final class DimensionValuesByteRefGroupingAggregatorFunction implements G
     @Override
     public void evaluateIntermediate(Block[] blocks, int offset, IntVector selected) {
         int positionCount = selected.getPositionCount();
-        boolean allSelected = positionCount == maxGroupId + 1;
+        boolean allSelected = positionCount > maxGroupId;
         if (allSelected) {
             for (int i = 0; i < selected.getPositionCount(); i++) {
-                if (selected.getInt(i) == i) {
+                if (selected.getInt(i) != i) {
                     allSelected = false;
                     break;
                 }
             }
         }
         if (allSelected) {
+            fillNullsUpTo(positionCount);
             blocks[offset] = builder.build();
             return;
         }

--- a/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/aggregate/DimensionValuesByteRefGroupingAggregatorFunctionTests.java
+++ b/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/aggregate/DimensionValuesByteRefGroupingAggregatorFunctionTests.java
@@ -77,14 +77,17 @@ public class DimensionValuesByteRefGroupingAggregatorFunctionTests extends Compu
             randomFrom(AggregatorMode.INITIAL, AggregatorMode.SINGLE),
             List.of(0)
         );
+        final List<BlockHash.GroupSpec> groupSpecs;
+        if (randomBoolean()) {
+            // Use IntBlockHash; reserves 0 for null values
+            groupSpecs = List.of(new BlockHash.GroupSpec(1, ElementType.INT));
+        } else {
+            // Use PackedValuesBlockHash; does not reserve 0 for null values
+            groupSpecs = List.of(new BlockHash.GroupSpec(1, ElementType.INT), new BlockHash.GroupSpec(1, ElementType.INT));
+        }
         HashAggregationOperator hashAggregationOperator = new HashAggregationOperator(
             List.of(aggregatorFactory),
-            () -> BlockHash.build(
-                List.of(new BlockHash.GroupSpec(1, ElementType.INT)),
-                driverContext.blockFactory(),
-                randomIntBetween(1, 1024),
-                randomBoolean()
-            ),
+            () -> BlockHash.build(groupSpecs, driverContext.blockFactory(), randomIntBetween(1, 1024), randomBoolean()),
             driverContext
         );
         List<Page> outputPages = new ArrayList<>();
@@ -99,7 +102,7 @@ public class DimensionValuesByteRefGroupingAggregatorFunctionTests extends Compu
         Map<Integer, List<BytesRef>> actualValues = new HashMap<>();
         for (Page out : outputPages) {
             IntBlock groups = out.getBlock(0);
-            BytesRefBlock valuesBlock = out.getBlock(1);
+            BytesRefBlock valuesBlock = out.getBlock(groupSpecs.size());
             for (int p = 0; p < out.getPositionCount(); p++) {
                 int group = groups.getInt(p);
                 int valueCount = valuesBlock.getValueCount(p);

